File: examples/counter/web-nextjs/pages/sse.ts
import {
  SseRequest,
  SseResponseVariantDone,
  SseResponseVariantChunk,
} from "shared_types/types/shared_types";

export async function* request({ url }: SseRequest) {
  const request = new Request(url);

  const response = await fetch(request);

  const reader = response.body.getReader();
  try {
    while (true) {
      const { done, value } = await reader.read();
      yield done
        ? new SseResponseVariantDone()
        : new SseResponseVariantChunk(Array.from(value));
      if (done) {
        break;
      }
    }
  } finally {
    reader.releaseLock();
  }
}

File: examples/counter/tui/src/sse.rs
use futures::{StreamExt, stream};

use reqwest::{Client, Method};
use shared::{
    http::{HttpError, Result},
    sse::{SseRequest, SseResponse},
};

pub async fn request(
    SseRequest { url }: &SseRequest,
) -> Result<impl futures::TryStream<Ok = SseResponse>> {
    let client = Client::new();
    let method = Method::from_bytes(b"GET").unwrap();

    let request = client
        .request(method, url)
        .build()
        .map_err(|e| HttpError::Url(e.to_string()))?;

    let response = client
        .execute(request)
        .await
        .map_err(|e| HttpError::Io(e.to_string()))?;

    let body = response.bytes_stream();

    Ok(Box::pin(stream::try_unfold(body, |mut body| async {
        body.next().await.map_or_else(
            || Ok(None),
            |chunk| match chunk {
                Ok(bytes) => {
                    let chunk = SseResponse::Chunk(bytes.to_vec());
                    Ok(Some((chunk, body)))
                }
                Err(e) => Err(HttpError::Json(format!(
                    "failed to read from http response; err = {e:?}"
                ))),
            },
        )
    })))
}
